### PR TITLE
Add resident field

### PR DESCRIPTION
**Pull request by jq1836**
@alvinjiang1 feel free to edit this comment, also, closed the original PR #58 

Update code to not depend on `Person` class is done in #73 

## Todo:
- [x] Update code to not depend on `Person` class
- [x] Fix tests

## Things done:
- Implemented 5 new classes: `Resident`, `Room`, `House`, `Gender` and `MatricNumber`.
- In each class, `VALIDATION_REGEX` has been updated to reflect the validation based on desired input as determined in our [User Guide](https://ay2223s1-cs2103t-w12-3.github.io/tp/UserGuide.html).
- Objects in the data classes have a **field of type String** that stores their respective information.
- Methods in `Resident` supports additional fields, compared to that of the old `Person` class.

## Documentation
- [ ] User Guide
- [ ] Developer Guide
- [X] Javadocs

### Sections updated under UG

- NIL

### Sections updated under DG

- NIL
@alvinjiang1 feel free to edit this comment

### Classes with Javadocs update

- `Resident`
- `Room`
- `Gender`
- `House`
- `MatricNumber`
- `Name`
- `Email`
- `Phone`

## Testing
**Testing to be done after shifting internal implementation of RC4HDB from using `Person` to using `Resident`.**

### CODE REVIEW

**Comment by jq1836** [comment]
# Todo:
- [ ] Update code to not depend on `Person` class
- [ ] Fix tests

## Update code to not depend on `Person` class
- [ ]

## Fix tests

## Proposed changes:
<--
1
-->
Feel free to add any ideas here.

File: src/main/java/seedu/rc4hdb/model/resident/fields/House.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        return test.matches(VALIDATION_REGEX);
+    }
+
+    /**
+     * Returns the house string as a full word.
+     * @return the house string as a full word.
+     */
+    public String asFullWord() {
+        switch (house) {
+        case ("D"):
```
Comment (nseah21) [inline src/main/java/seedu/rc4hdb/model/resident/fields/House.java:43]: Some formatting issues here. Also, should we omit the parentheses for the case clauses? I believe `case "D":` would suffice.

Comment (nseah21) [reply src/main/java/seedu/rc4hdb/model/resident/fields/House.java:43]: The same comments apply to the other cases for the rest of the switch statement.

**Review by nealetham — APPROVED** [review]
LGTM! Very comprehensive testing done by @Jordan
Great work all!

File: src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java [hunk 2]
```diff
@@ -1,142 +1,221 @@
...
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ROOM_DESC_BOB
+                + GENDER_DESC_AMY + GENDER_DESC_BOB + HOUSE_DESC_BOB + MATRIC_NUMBER_DESC_BOB + TAG_DESC_FRIEND,
+                new AddCommand(expectedResident));
+
+        // multiple houses - last room accepted
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ROOM_DESC_BOB
+                + GENDER_DESC_BOB + HOUSE_DESC_AMY + HOUSE_DESC_BOB + MATRIC_NUMBER_DESC_BOB + TAG_DESC_FRIEND,
+                new AddCommand(expectedResident));
+
+        // multiple matric numbers - last room accepted
```
Comment (nseah21) [inline src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java:101]: Some issues with the comments here. Might want to update "last room accepted" to reflect the corresponding field. 😄

File: src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java [hunk 3]
```diff
@@ -1,142 +1,221 @@
...
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ROOM_DESC_AMY
+                + ROOM_DESC_BOB + GENDER_DESC_BOB + HOUSE_DESC_BOB + MATRIC_NUMBER_DESC_BOB + TAG_DESC_FRIEND,
+                new AddCommand(expectedResident));
+
+        // multiple gender - last room accepted
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ROOM_DESC_BOB
+                + GENDER_DESC_AMY + GENDER_DESC_BOB + HOUSE_DESC_BOB + MATRIC_NUMBER_DESC_BOB + TAG_DESC_FRIEND,
+                new AddCommand(expectedResident));
+
+        // multiple houses - last room accepted
```
Comment (nseah21) [inline src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java:96]:
```suggestion
        // multiple houses - last house accepted
```

File: src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java [hunk 4]
```diff
@@ -1,142 +1,221 @@
...
 
-        // multiple addresses - last address accepted
-        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_AMY
-                + ADDRESS_DESC_BOB + TAG_DESC_FRIEND, new AddCommand(expectedPerson));
+        // multiple rooms - last room accepted
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ROOM_DESC_AMY
+                + ROOM_DESC_BOB + GENDER_DESC_BOB + HOUSE_DESC_BOB + MATRIC_NUMBER_DESC_BOB + TAG_DESC_FRIEND,
+                new AddCommand(expectedResident));
+
+        // multiple gender - last room accepted
```
Comment (nseah21) [inline src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java:91]:
```suggestion
        // multiple gender - last gender accepted
```

**Review by nseah21 — APPROVED** [review]
LGTM. Some minor nitpicks on code quality but overall good job!

File: src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java (hunk 2 above)
Comment (jq1836) [reply src/test/java/seedu/rc4hdb/logic/parser/AddCommandParserTest.java:101]: oh right hahaha

**Review by alvinjiang1 — APPROVED** [review]
LGTM👌

I will check on UG/DG again, but it should have been updated as of to include the changes from this PR already so all's good :)

**Pull request merged**